// @stripe/firestore-stripe-payments/lib/subscription.js
/*
 * Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { collection, doc, getDoc, getDocs, getFirestore, onSnapshot, query, where, } from "@firebase/firestore";
import { StripePaymentsError } from "./init";
import { getCurrentUser, getCurrentUserSync } from "./user";
import { checkNonEmptyArray, checkNonEmptyString } from "./utils";
/**
 * Retrieves an existing Stripe subscription for the currently signed in user from the database.
 *
 * @param payments - A valid {@link StripePayments} object.
 * @param subscriptionId - ID of the subscription to retrieve.
 * @returns Resolves with a Subscription object if found. Rejects if the specified subscription ID
 *  does not exist, or if the user is not signed in.
 */
export function getCurrentUserSubscription(payments, subscriptionId) {
    checkNonEmptyString(subscriptionId, "subscriptionId must be a non-empty string.");
    return getCurrentUser(payments).then((uid) => {
        const dao = getOrInitSubscriptionDAO(payments);
        return dao.getSubscription(uid, subscriptionId);
    });
}
/**
 * Retrieves existing Stripe subscriptions for the currently signed in user from the database.
 *
 * @param payments - A valid {@link StripePayments} object.
 * @param options - A set of options to customize the behavior.
 * @returns Resolves with an array of Stripe subscriptions. May be empty.
 */
export function getCurrentUserSubscriptions(payments, options) {
    const queryOptions = {};
    if (typeof (options === null || options === void 0 ? void 0 : options.status) !== "undefined") {
        queryOptions.status = getStatusAsArray(options.status);
    }
    return getCurrentUser(payments).then((uid) => {
        const dao = getOrInitSubscriptionDAO(payments);
        return dao.getSubscriptions(uid, queryOptions);
    });
}
/**
 * Registers a listener to receive subscription update events for the currently signed in
 * user. If the user is not signed in throws an `unauthenticated` error, and no listener is
 * registered.
 *
 * Upon successful registration, the `onUpdate` callback will fire once with
 * the current state of all the subscriptions. From then onwards, each update to a subscription
 * will fire the `onUpdate` callback with the latest state of the subscriptions.
 *
 * @param payments - A valid {@link StripePayments} object.
 * @param onUpdate - A callback that will fire whenever the current user's subscriptions
 *   are updated.
 * @param onError - A callback that will fire whenever an error occurs while listening to
 *   subscription updates.
 * @returns A function that can be called to cancel and unregister the listener.
 */
export function onCurrentUserSubscriptionUpdate(payments, onUpdate, onError) {
    const uid = getCurrentUserSync(payments);
    const dao = getOrInitSubscriptionDAO(payments);
    return dao.onSubscriptionUpdate(uid, onUpdate, onError);
}
function getStatusAsArray(status) {
    if (typeof status === "string") {
        return [status];
    }
    checkNonEmptyArray(status, "status must be a non-empty array.");
    return status;
}
const SUBSCRIPTION_CONVERTER = {
    toFirestore: () => {
        throw new Error("Not implemented for readonly Subscription type.");
    },
    fromFirestore: (snapshot) => {
        var _a, _b, _c;
        const data = snapshot.data();
        const refs = data.prices;
        const prices = refs.map((priceRef) => {
            return {
                product: priceRef.parent.parent.id,
                price: priceRef.id,
            };
        });
        return {
            cancel_at: toNullableUTCDateString(data.cancel_at),
            cancel_at_period_end: data.cancel_at_period_end,
            canceled_at: toNullableUTCDateString(data.canceled_at),
            created: toUTCDateString(data.created),
            current_period_start: toUTCDateString(data.current_period_start),
            current_period_end: toUTCDateString(data.current_period_end),
            ended_at: toNullableUTCDateString(data.ended_at),
            id: snapshot.id,
            metadata: (_a = data.metadata) !== null && _a !== void 0 ? _a : {},
            price: data.price.id,
            prices,
            product: data.product.id,
            quantity: (_b = data.quantity) !== null && _b !== void 0 ? _b : null,
            role: (_c = data.role) !== null && _c !== void 0 ? _c : null,
            status: data.status,
            stripe_link: data.stripeLink,
            trial_end: toNullableUTCDateString(data.trial_end),
            trial_start: toNullableUTCDateString(data.trial_start),
            uid: snapshot.ref.parent.parent.id,
        };
    },
};
const SUBSCRIPTIONS_COLLECTION = "subscriptions";
function toNullableUTCDateString(timestamp) {
    if (timestamp === null) {
        return null;
    }
    return toUTCDateString(timestamp);
}
function toUTCDateString(timestamp) {
    return timestamp.toDate().toUTCString();
}
class FirestoreSubscriptionDAO {
    constructor(app, customersCollection) {
        this.customersCollection = customersCollection;
        this.firestore = getFirestore(app);
    }
    getSubscription(uid, subscriptionId) {
        return __awaiter(this, void 0, void 0, function* () {
            const snap = yield this.getSubscriptionSnapshotIfExists(uid, subscriptionId);
            return snap.data();
        });
    }
    getSubscriptions(uid, options) {
        return __awaiter(this, void 0, void 0, function* () {
            const querySnap = yield this.getSubscriptionSnapshots(uid, options === null || options === void 0 ? void 0 : options.status);
            const subscriptions = [];
            querySnap.forEach((snap) => {
                subscriptions.push(snap.data());
            });
            return subscriptions;
        });
    }
    onSubscriptionUpdate(uid, onUpdate, onError) {
        const subscriptions = collection(this.firestore, this.customersCollection, uid, SUBSCRIPTIONS_COLLECTION).withConverter(SUBSCRIPTION_CONVERTER);
        return onSnapshot(subscriptions, (querySnap) => {
            const snapshot = {
                subscriptions: [],
                changes: [],
                size: querySnap.size,
                empty: querySnap.empty,
            };
            querySnap.forEach((snap) => {
                snapshot.subscriptions.push(snap.data());
            });
            querySnap
                .docChanges()
                .forEach((change) => {
                snapshot.changes.push({
                    type: change.type,
                    subscription: change.doc.data(),
                });
            });
            onUpdate(snapshot);
        }, (err) => {
            if (onError) {
                const arg = new StripePaymentsError("internal", `Error while listening to database updates: ${err.message}`, err);
                onError(arg);
            }
        });
    }
    getSubscriptionSnapshotIfExists(uid, subscriptionId) {
        return __awaiter(this, void 0, void 0, function* () {
            const subscriptionRef = doc(this.firestore, this.customersCollection, uid, SUBSCRIPTIONS_COLLECTION, subscriptionId).withConverter(SUBSCRIPTION_CONVERTER);
            const snapshot = yield this.queryFirestore(() => getDoc(subscriptionRef));
            if (!snapshot.exists()) {
                throw new StripePaymentsError("not-found", `No subscription found with the ID: ${subscriptionId} for user: ${uid}`);
            }
            return snapshot;
        });
    }
    getSubscriptionSnapshots(uid, status) {
        return __awaiter(this, void 0, void 0, function* () {
            let subscriptionsQuery = collection(this.firestore, this.customersCollection, uid, SUBSCRIPTIONS_COLLECTION).withConverter(SUBSCRIPTION_CONVERTER);
            if (status) {
                subscriptionsQuery = query(subscriptionsQuery, where("status", "in", status));
            }
            return yield this.queryFirestore(() => getDocs(subscriptionsQuery));
        });
    }
    queryFirestore(fn) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                return yield fn();
            }
            catch (error) {
                throw new StripePaymentsError("internal", "Unexpected error while querying Firestore", error);
            }
        });
    }
}
const SUBSCRIPTION_DAO_KEY = "subscription-dao";
function getOrInitSubscriptionDAO(payments) {
    let dao = payments.getComponent(SUBSCRIPTION_DAO_KEY);
    if (!dao) {
        dao = new FirestoreSubscriptionDAO(payments.app, payments.customersCollection);
        setSubscriptionDAO(payments, dao);
    }
    return dao;
}
/**
 * Internal API for registering a {@link SubscriptionDAO} instance with {@link StripePayments}.
 * Exported for testing.
 *
 * @internal
 */
export function setSubscriptionDAO(payments, dao) {
    payments.setComponent(SUBSCRIPTION_DAO_KEY, dao);
}
